src/couch_views/src/couch_views_jobs.erl

```diff
@@ -84,21 +91,35 @@ wait_for_job(JobId, UpdateSeq) ->
     end.
 
 
-wait_for_job(JobId, Subscription, UpdateSeq) ->
+wait_for_job(JobId, Subscription, DDocId, UpdateSeq) ->
     case wait(Subscription) of
+        {not_found, not_found} ->
+            erlang:error(index_not_found);
         {error, Error} ->
             erlang:error(Error);
+        {finished, #{<<"error">> := <<"ddoc_deleted">>} = Data} ->
+            case maps:get(<<"ddoc_id">>, Data) of
+                DDocId ->
+                    couch_jobs:remove(undefined, ?INDEX_JOB_TYPE, JobId),
+                    erlang:error({ddoc_deleted, maps:get(<<"reason">>, Data)});
+                _OtherDocId ->
+                    % A different design doc wiht the same signature
+                    % was deleted. Resubmit this job which will overwrite
+                    % the ddoc_id in the job.
+                    retry
+            end;
         {finished, #{<<"error">> := Error, <<"reason">> := Reason}} ->
+            couch_jobs:remove(undefined, ?INDEX_JOB_TYPE, JobId),
```

Comment:
This will work great for `sig_changed` and `db_deleted` but we also fail for any exceptions [1] after we exceed the max retries number or even right away if users picked a large row limit config and they get "tx too long" error. Then current subscribers will get the error but if there is no active request waiting for it and they retry the view later after it finished if we'd want to preserve the failed job?

Maybe we don't because we have some subtle behavior in [2] where if they query the job later, and don't find a `<<"view_seq">>` since it will be an error, they'll get a `retry` will re-add the job and  worker will re-build it again. 

[1] https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/couch_views/src/couch_views_indexer.erl#L116

[2] https://github.com/apache/couchdb/blob/prototype/fdb-layer/src/couch_views/src/couch_views_jobs.erl#L82

Reply:
I agree that the retry is going to retry eventually. This is just to cleanup when a view errors out so we're not leaving orphaned job descriptions everywhere.

As to error conditions in the retry logic anything that comes from there would either be a bug in our code or a transient error that we should eventually be able to get past so I don't think we should be changing things to never remove that job info and always return the same error. (which is not what we're currently doing as you point out).